pkg/reconciler/pipelinerun/pipelinerun_test.go

```diff
@@ -3895,21 +3617,60 @@ func TestReconcile_CloudEvents(t *testing.T) {
 		t.Errorf("Expected PipelineRun status to include the TaskRun status items that can run immediately: %v", reconciledRun.Status.TaskRuns)
 	}
 
-	wantEvents := []string{
-		"Normal Started",
-		"Normal Running Tasks Completed: 0",
-	}
-	err = checkEvents(t, testAssets.Recorder, "reconcile-cloud-events", wantEvents)
-	if !(err == nil) {
-		t.Errorf(err.Error())
-	}
 	wantCloudEvents := []string{
 		`(?s)dev.tekton.event.pipelinerun.started.v1.*test-pipelinerun`,
 		`(?s)dev.tekton.event.pipelinerun.running.v1.*test-pipelinerun`,
 	}
 	ceClient := clients.CloudEvents.(cloudevent.FakeClient)
-	err = checkCloudEvents(t, &ceClient, "reconcile-cloud-events", wantCloudEvents)
+	err := checkCloudEvents(t, &ceClient, "reconcile-cloud-events", wantCloudEvents)
 	if !(err == nil) {
 		t.Errorf(err.Error())
 	}
 }
+
+// NewPipelineRunTest returns PipelineRunTest with a new PipelineRun controller created with specified state through data
+// This PipelineRunTest can be reused for multiple PipelineRuns by calling reconcileRun for each pipelineRun
+func NewPipelineRunTest(data test.Data, t *testing.T) *PipelineRunTest {
```

Comment:
i'm now just nitpicking, but i'd put the functions that are related to PipelineRunTest near the declaration of the struct. dont wanna block the PR tho!

Reply:
thanks @bobcatfish I was planning on moving all helper functions including:

*  `PipelineRunTest` struct declaration and function related to it
* getPipelineRunController
* conditionCheckFromTaskRun
* checkEvents
* eventFromChannel

etc, to a separate file `/pkg/reconciler/pipelinerun/pipelinerun_helpers_test.go` as a follow up PR. `pkg/reconciler/pipelinerun/pipelinerun_test.go` is a combination of helper utilities and actual unit tests and helper functions are spread across this file 😞 Will be easier to introduce more tests and/or reviewing them with such separation.

Reply:
okay great!

A couple of thoughts:

* Somewhere in the top level [test](https://github.com/tektoncd/pipeline/tree/master/test) directory might be a good place to go, since that is where we are already putting our test libs there (might even make sense to work it into https://github.com/tektoncd/pipeline/blob/master/test/controller.go)
* I want to caution against naming and calling code "helpers" - this is totally a common practice but I think it's useful to get out of the habit, for a couple of reasons:
  * People tend to use this terminology about libs used for test code - sometimes they use it for the code under tests as well but i think that's less common - and it seems almost like it's putting the code down, i.e., "i need this code to help me" whereas I think that well factored test code is just as important as any other well factored code, so libraries for test code are just as important as other code < rant over >
  * The other (probably bigger and less my personal opinion) reason is that "helper" is not a very specific name - good naming and code organization requires thinking about what the code's responsibility is and naming and organizing it based on that; using names like "helper" "manager" "library" "util" is a short cut that prevents us from doing that. More on this at https://dave.cheney.net/2019/01/08/avoid-package-names-like-base-util-or-common and https://blog.golang.org/package-names in the section "bad package names":

> Avoid meaningless package names. Packages named util, common, or misc provide clients with no sense of what the package contains. This makes it harder for clients to use the package and makes it harder for maintainers to keep the package focused. Over time, they accumulate dependencies that can make compilation significantly and unnecessarily slower, especially in large programs. And since such package names are generic, they are more likely to collide with other packages imported by client code, forcing clients to invent names to distinguish them.